clustering.py

```diff
@@ -244,3 +246,220 @@ def sort_labels(label_dict, cluster_labels, ts_to_labels):
         ordered_labels[i] = system_labels[elt]
 
     return ordered_labels, ordered_cluster_labels, ordered_ts_labels
+
+def kmeans_constrained(data, label_dict, ts_to_labels):
+    """Runs k-means with constraints and uses k-means++ initialization.
+
+    Args:
+        data: An np array where each row is a time series and each
+            column is a time.
+        label_dict: A dictionary where the keys are labels and the
+            values are the indexes of the labels in data.
+        ts_to_labels: An array where each row is a timeSeries and each
+            column is a label.
+
+    Returns:
+        An np array where the ith element is the cluster the ith time
+        series was placed in.
+    """
+    must_link, can_not_link = make_constraints(ts_to_labels)
+    num_clusters = (len(data) // 20) + 6
+    clusters_distances = []
+
+    for run_num in range(5):
+        centroids = k_means_init(data, num_clusters, run_num)
+
+        while True:
+            old_centroids = centroids.copy()
+
+            clusters, ts_to_cluster = update_clusters(data, centroids,
+                                                      must_link, can_not_link)
+            valid_clusters = update_centroids(data, clusters, centroids)
+
+            if not valid_clusters:
+                break
+
+            if np.array_equal(old_centroids, centroids):
+                assignment = [v for k, v in sorted(ts_to_cluster.items(),
+                                                   key=lambda item: item[0])]
+                center_dist = 0
+                for index, cluster in enumerate(assignment):
+                    center_dist += distance.euclidean(data[index],
+                                                      centroids[cluster])
+                clusters_distances.append([center_dist, assignment, run_num])
+                break
+
+    clusters_distances.sort()
+    return np.array(clusters_distances[0][1])
+
+def update_clusters(data, centroids, must_link, can_not_link):
+    """Updates the cluster assignments based on the centroids.
+
+    Args:
+        data: An np array where each row is a time series and each
+            column is a time.
+        centroids: A list of the centroids.
+        must_link: A dictionary mapping time series that must link.
+        can_not_link: A dictionary mapping time series that can't link.
+
+    Returns:
+        clusters: A list where the ith element is a list of the indexes
+            of the elements in the ith cluster.
+        ts_to_cluster: A dictionary mapping each time series to its
+            cluster assignment.
+    """
+    distances = pairwise_distances(data, centroids)
+    clusters = [[] for i in range(len(centroids))]
+    ts_to_cluster = {}
+
+    for ts_index, dist in enumerate(distances):
+        options = np.argsort(dist)
+        for option in options:
```

Comment:
Could you comment more about what's going on here? Reading the documentation for argsort, it looks like options are the indices in order if dist were sorted, so this inner for loop assigns the timeseries to the cluster with the shortest distance that does not violate the constraints function.

Reply:
Yes, that's exactly what happens. I'll add that to the comment.